File: Homework5/Program/Classes/Parser/RPNParser.java
package Homework5.Program.Classes.Parser;

import java.util.Stack;

import Homework5.Program.Classes.Calc.Calc;

public class RPNParser {
    private static Stack<Float> stack = new Stack<>();

    public static float parse(Stack<String> tokens) {
        while (!tokens.isEmpty()) {
            String token = tokens.pop();
            try {
                stack.push(Float.parseFloat(token));
            } catch (Exception e) {
                float b = stack.pop();
                float a = stack.pop();
                switch (token.charAt(0)) {
                    case '+':
                        stack.push(Calc.sum(a, b));
                        break;
                    case '-':
                        stack.push(Calc.sub(a, b));
                        break;
                    case '*':
                        stack.push(Calc.mul(a, b));
                        break;
                    case '/':
                        stack.push(Calc.div(a, b));
                        break;
                }
            }
        }
        return stack.pop();
    }
}

File: Homework7/model/RPNParser.py
import Ops as op
from collections import deque

class RPN_Parser:
    ops = ['(', ')', '-', '+', '/', '*', '^']
    stack = deque()

    def parse(self, rpn:deque):
        stack = self.stack
        ops = self.ops
        while len(rpn)>0:
            token = rpn.popleft()
            if token in ops:
                b = float(stack.pop())
                a = float(stack.pop())
                if token == '+':
                    stack.append(op.sum(a, b))
                elif token == "-":
                    stack.append(op.sub(a, b))
                elif token == '*':
                    stack.append(op.mul(a, b))
                elif token == '/':
                    stack.append(op.div(a, b))
            else:
                stack.append(token)
        return stack.pop()


